Refactored close procedures to dispatch Errors for all components

Drizzt1991 opened the pull request:
Basically now it will always raise appropriate exceptions for any public calls after closure. Added ClientClosed and ServerClosed exceptions. Maybe they are a little to grained, your thoughts?

Now we can write:
```python
@asyncio.coroutine
def do_stuff(connection):
    queue = yield from declare_queue(connection)
    consumer = yield from queue.consume(lambda x: print(x), no_ack=True)
    yield from asyncio.sleep(1)
    queue.publish(some_message)

connection = yield from connect()
while True:
    try:
        yield from consume(connection)
    except asynqp.exceptions.AlreadyClosed:
        connection = yield from reconnect()
```
and it will raise no matter where connection is lost.

Depends on #54. Please review that firts. I kinda split them in a way )

Drizzt1991 commented:
Rebased

Drizzt1991 commented:
Bump. I would really appreciate feedback on this, as it's a stopper for #34

benjamin-hodgson commented:
Apologies, I haven't had time to review this. I'll try and do it ASAP

benjamin-hodgson commented on src/asynqp/protocol.py:
```diff
@@ -58,11 +60,17 @@ def connection_lost(self, exc):
+        self.close()
+
+    def close(self):
+        assert not self._closed, "Why do we close it 2-ce?"
```
Is this _possible_? If a user could provoke this condition by (ab)using the API, we should explicitly throw an exception.

Drizzt1991 replied:
I think it's not possible. We didn't specify protocol as part of public API in docs.

benjamin-hodgson commented on src/asynqp/routing.py:
```diff
@@ -21,8 +18,15 @@ def remove_handler(self, channel_id):
+                                        spec.ConnectionClose)):
+            # Notify all connection channels about it
+            for handler in self.handlers.values():
+                handler(frame)
```
It looks like the idea is to always dispatch `ConnectionClose` to all handlers. Does the existing 'poison pill' infrastructure not serve your purpose?

Drizzt1991 replied:
Will do it using the poison pill

benjamin-hodgson commented on src/asynqp/routing.py:
```diff
@@ -33,22 +37,27 @@ class Sender(object):
+        """ Connection/Channel was closed. All subsequent requests should
+            raise an error
+        """
+        self._connection_exc = exc
```
Doesn't feel like this belongs in the `Sender`

Drizzt1991 replied:
The usecase I want to cover is like so:
```
exchange = yield from create_exchange()
while True:
    try:
        event = yield from wait_for_event()
        exchange.publish(event)
    except AMQPError:
        exchange = reconnect_and_create_exchange()
```
In other words, those methods don't raise errors on disconnect:
* publish
* ack/reject

Drizzt1991 replied:
I will remove it from this PR, and create a separate issue, so we can return to it later.

benjamin-hodgson commented on src/asynqp/protocol.py:
```diff
@@ -8,10 +8,12 @@
 
 class AMQP(asyncio.Protocol):
     def __init__(self, dispatcher, loop):
+        self._loop = loop
```
Is this used?

Drizzt1991 replied:
maybe a rebase issue

Drizzt1991 commented:
Squashed and cleaned code. Removed some unneeded changes, so they don't bother in review.
The idea to use PoisonPill for closing was quite helpful, I like this one more.

Drizzt1991 commented:
bump again
